src/ride/water/submarine_ride.c

```diff
@@ -64,3 +65,349 @@ void vehicle_visual_submarine(int x, int imageDirection, int y, int z, rct_vehic
+	SPR_SUBMARINE_RIDE_25_DEG_UP_TO_FLAT_SE_NW,
+
+	SPR_SUBMARINE_RIDE_25_DEG_UP_SW_NE,
+	SPR_SUBMARINE_RIDE_25_DEG_UP_NW_SE,
+	SPR_SUBMARINE_RIDE_25_DEG_UP_NE_SW,
+	SPR_SUBMARINE_RIDE_25_DEG_UP_SE_NW,
+
+	SPR_SUBMARINE_RIDE_FLAT_QUARTER_TURN_1_TILE_SW_NW = 16896,
+	SPR_SUBMARINE_RIDE_FLAT_QUARTER_TURN_1_TILE_NW_NE = 16897,
+	SPR_SUBMARINE_RIDE_FLAT_QUARTER_TURN_1_TILE_NE_SE = 16898,
+	SPR_SUBMARINE_RIDE_FLAT_QUARTER_TURN_1_TILE_SE_SW = 16899,
+};
+
+static void submarine_ride_paint_track_station(uint8 rideIndex, uint8 trackSequence, uint8 direction, int height, rct_map_element * mapElement)
+{
+	rct_xy16 position = {RCT2_GLOBAL(0x009DE56A, sint16), RCT2_GLOBAL(0x009DE56E, sint16)};
+	rct_ride * ride = get_ride(rideIndex);
+	const rct_ride_entrance_definition * entranceStyle = &RideEntranceDefinitions[ride->entrance_style];
+	int heightLower = height - 16;
+	uint32 imageId;
+	bool hasFence;
+
+	if (direction & 1) {
+		imageId = SPR_SUBMARINE_RIDE_FLAT_SE_NW | RCT2_GLOBAL(0x00F44198, uint32);
+		sub_98197C(imageId, 0, 0, 20, 32, 3, heightLower, 6, 0, heightLower, get_current_rotation());
+		paint_util_push_tunnel_right(height, TUNNEL_6);
+
+		hasFence = track_paint_util_has_fence(EDGE_NE, position, mapElement, ride, get_current_rotation());
+		imageId = (hasFence ? SPR_STATION_PIER_EDGE_NE_FENCED : SPR_STATION_PIER_EDGE_NE) | RCT2_GLOBAL(0x00F4419C, uint32);
+		sub_98197C(imageId, 0, 0, 6, 32, 1, height, 2, 0, height, get_current_rotation());
+		track_paint_util_draw_station_covers(EDGE_NE, hasFence, entranceStyle, direction, height);
+
+		imageId = SPR_STATION_PIER_EDGE_SW | RCT2_GLOBAL(0x00F4419C, uint32);
+		sub_98196C(imageId, 24, 0, 8, 32, 1, height, get_current_rotation());
+
+		if (RCT2_GLOBAL(0x0141E9DB, uint8) & 3) {
+			hasFence = track_paint_util_has_fence(EDGE_SW, position, mapElement, ride, get_current_rotation());
+			if (hasFence) {
+				imageId = SPR_STATION_PIER_FENCE_SW | RCT2_GLOBAL(0x00F4419C, uint32);
+				sub_98196C(imageId, 31, 0, 1, 32, 7, height + 2, get_current_rotation());
+			}
+			track_paint_util_draw_station_covers(EDGE_SW, hasFence, entranceStyle, direction, height);
+		}
+
+	} else {
+		imageId = SPR_SUBMARINE_RIDE_FLAT_NE_SW | RCT2_GLOBAL(0x00F44198, uint32);
+		sub_98197C(imageId, 0, 0, 32, 20, 3, heightLower, 0, 6, heightLower, get_current_rotation());
+		paint_util_push_tunnel_left(height, TUNNEL_6);
+
+		hasFence = track_paint_util_has_fence(EDGE_NW, position, mapElement, ride, get_current_rotation());
+		imageId = (hasFence ? SPR_STATION_PIER_EDGE_NW_FENCED : SPR_STATION_PIER_EDGE_NW) | RCT2_GLOBAL(0x00F4419C, uint32);
+		sub_98197C(imageId, 0, 0, 32, 6, 1, height, 0, 2, height, get_current_rotation());
+		track_paint_util_draw_station_covers(EDGE_NW, hasFence, entranceStyle, direction, height);
+
+		imageId = SPR_STATION_PIER_EDGE_SE | RCT2_GLOBAL(0x00F4419C, uint32);
+		sub_98196C(imageId, 0, 24, 32, 8, 1, height, get_current_rotation());
+
+		if (RCT2_GLOBAL(0x0141E9DB, uint8) & 3) {
```

Comment:
I don't think that this is how the station pieces should be drawn. This doesn't make a difference for water rides but I feel we should stick to the same pattern. In my opinion when underground we should draw just like above ground but without the station covers. What happens in vanilla is that the station is drawn missing the front fence and drawing half of the station cover. This is clearly not quite right.